File: LeetCode/Kai/63UniquePathsII.java
public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        if(m==1 && n==1) return obstacleGrid[0][0]==1 ? 0 : 1;  //corner case
        
        int[][] dp = new int[m][n];
        for(int i=0; i<m; i++) Arrays.fill(dp[i], 0);
        //base
        for(int i=0; i<m; i++) {
            if(obstacleGrid[i][0] == 1) break;
            dp[i][0] = 1;
        }
        for(int i=0; i<n; i++) {
            if(obstacleGrid[0][i] == 1) break;
            dp[0][i] = 1;
        }
        
        for(int i=1; i<m; i++) {
            for(int j=1; j<n; j++) {
                if(obstacleGrid[i][j]==1) dp[i][j] = 0;
                else dp[i][j] = dp[i][j-1]+dp[i-1][j];
            }
        }
        return dp[m-1][n-1];
    }
    
}

File: LeetCode/Lei/63UniquePathsII.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        vector<vector<int>> u = obstacleGrid;
        int m = int(obstacleGrid.size());
        int n = int(obstacleGrid[0].size());
        for(int i=0; i<int(obstacleGrid.size()); i++)
        {
          for(int j=0; j<int(obstacleGrid[0].size()); j++)
          {
            if(obstacleGrid[i][j]==1)
            {
              u[i][j]=0;
            }
            else if(i>0&&j>0)
            {
              u[i][j]=u[i-1][j]+u[i][j-1];
            }
            else if(i>0)
            {
              u[i][j]=u[i-1][j];
            }
            else if(j>0)
            {
              u[i][j]=u[i][j-1];
            }
            else{
              u[i][j]=1;
            }
          }
        }
    return u[m-1][n-1];
    }
};
